File: go/tests/integration_test.go
package tests

import (
	"context"
	"path/filepath"
	"testing"

	"github.com/shopspring/decimal"

	"github.com/DexwoxBusiness/dexcost-go/core"
	"github.com/DexwoxBusiness/dexcost-go/pricing"
	"github.com/DexwoxBusiness/dexcost-go/transport"
)

// newTestTracker creates a Tracker backed by a temporary SQLiteBuffer.
func newTestTracker(t *testing.T, dbName string, opts ...func(*core.TrackerOptions)) *core.Tracker {
	t.Helper()
	dir := t.TempDir()
	dbPath := filepath.Join(dir, dbName)
	buf, err := transport.NewSQLiteBuffer(dbPath)
	if err != nil {
		t.Fatalf("create buffer: %v", err)
	}
	topts := core.TrackerOptions{Buffer: buf}
	for _, o := range opts {
		o(&topts)
	}
	tracker, err := core.NewTracker(topts)
	if err != nil {
		buf.Close()
		t.Fatalf("create tracker: %v", err)
	}
	t.Cleanup(func() { tracker.Close() })
	return tracker
}

// TestEndToEnd_FullWorkflow exercises the complete SDK workflow:
// init -> start task -> record LLM call -> record non-LLM cost -> mark retry -> end task
// Then queries the buffer and verifies all aggregated fields.
func TestEndToEnd_FullWorkflow(t *testing.T) {
	// Create tracker with pricing engine and rate registry.
	rates := pricing.NewRateRegistry()
	rates.Register("google_maps", "request", decimal.RequireFromString("0.005"))

	tracker := newTestTracker(t, "integration.db", func(o *core.TrackerOptions) {
		o.Rates = rates
	})

	// 1. Start a task with customer and project.
	ctx, tt := tracker.StartTask(context.Background(), "resolve_ticket",
		core.WithCustomer("acme-corp"),
		core.WithProject("support-q1"),
	)

	// Verify task was created correctly.
	if tt.Task.TaskType != "resolve_ticket" {
		t.Errorf("expected resolve_ticket, got %s", tt.Task.TaskType)
	}
	if tt.Task.CustomerID != "acme-corp" {
		t.Errorf("expected acme-corp, got %s", tt.Task.CustomerID)
	}
	if tt.Task.ProjectID != "support-q1" {
		t.Errorf("expected support-q1, got %s", tt.Task.ProjectID)
	}
	if tt.Task.Status != core.TaskStatusRunning {
		t.Errorf("expected running, got %s", tt.Task.Status)
	}

	// 2. Record an LLM call (auto-priced from bundled data).
	err := tt.RecordLLMCall("openai", "gpt-4o", 1000, 500)
	if err != nil {
		t.Fatalf("record LLM call: %v", err)
	}

	// 3. Record a non-LLM cost via rate registry.
	err = tt.RecordUsage("google_maps", 3)
	if err != nil {
		t.Fatalf("record usage: %v", err)
	}

	// 4. Record a direct cost.
	err = tt.RecordCost("stripe_api", decimal.RequireFromString("0.025"))
	if err != nil {
		t.Fatalf("record cost: %v", err)
	}

	// 5. Mark a retry.
	retryCost := decimal.RequireFromString("0.003")
	err = tt.MarkRetry("rate_limit", core.WithRetryCost(retryCost))
	if err != nil {
		t.Fatalf("mark retry: %v", err)
	}

	// 6. End the task.
	err = tt.End(core.TaskStatusSuccess)
	if err != nil {
		t.Fatalf("end task: %v", err)
	}

	// 7. Verify aggregated fields.
	stored, err := tracker.Buffer().GetTask(tt.Task.TaskID.String())
	if err != nil {
		t.Fatalf("get task: %v", err)
	}
	if stored == nil {
		t.Fatal("expected stored task")
	}

	// Status should be success.
	if stored.Status != core.TaskStatusSuccess {
		t.Errorf("expected success, got %s", stored.Status)
	}

	// EndedAt should be set.
	if stored.EndedAt == nil {
		t.Error("expected ended_at to be set")
	}

	// LLM cost should be non-zero (auto-priced gpt-4o).
	if stored.LLMCostUSD.IsZero() {
		t.Error("expected non-zero LLM cost (auto-priced gpt-4o)")
	}

	// External cost = google_maps (3 * 0.005 = 0.015) + stripe_api (0.025) = 0.04
	expectedExternal := decimal.RequireFromString("0.04")
	if !stored.ExternalCostUSD.Equal(expectedExternal) {
		t.Errorf("expected external_cost=%s, got %s", expectedExternal, stored.ExternalCostUSD)
	}

	// Total cost = LLM + external + compute.
	expectedTotal := stored.LLMCostUSD.Add(stored.ExternalCostUSD).Add(stored.ComputeCostUSD)
	if !stored.TotalCostUSD.Equal(expectedTotal) {
		t.Errorf("expected total_cost=%s, got %s", expectedTotal, stored.TotalCostUSD)
	}

	// Retry count and cost.
	if stored.RetryCount != 1 {
		t.Errorf("expected retry_count=1, got %d", stored.RetryCount)
	}
	if !stored.RetryCostUSD.Equal(retryCost) {
		t.Errorf("expected retry_cost=%s, got %s", retryCost, stored.RetryCostUSD)
	}

	// Token totals from LLM call.
	if stored.TotalInputTokens != 1000 {
		t.Errorf("expected input_tokens=1000, got %d", stored.TotalInputTokens)
	}
	if stored.TotalOutputTokens != 500 {
		t.Errorf("expected output_tokens=500, got %d", stored.TotalOutputTokens)
	}

	// Failure count should be 0 for success.
	if stored.FailureCount != 0 {
		t.Errorf("expected failure_count=0, got %d", stored.FailureCount)
	}

	// 8. Verify event count and types.
	events, err := tracker.Buffer().QueryEvents(tt.Task.TaskID.String())
	if err != nil {
		t.Fatalf("query events: %v", err)
	}
	if len(events) != 4 {
		t.Fatalf("expected 4 events, got %d", len(events))
	}

	// Verify event types.
	typeCount := make(map[core.EventType]int)
	for _, e := range events {
		typeCount[e.EventType]++
	}
	if typeCount[core.EventTypeLLMCall] != 1 {
		t.Errorf("expected 1 llm_call, got %d", typeCount[core.EventTypeLLMCall])
	}
	if typeCount[core.EventTypeExternalCost] != 2 {
		t.Errorf("expected 2 external_cost, got %d", typeCount[core.EventTypeExternalCost])
	}
	if typeCount[core.EventTypeRetryMarker] != 1 {
		t.Errorf("expected 1 retry_marker, got %d", typeCount[core.EventTypeRetryMarker])
	}

	// Verify context had the task.
	ctxTask := core.GetCurrentTask(ctx)
	if ctxTask == nil {
		t.Error("expected task in context")
	}
}

// TestEndToEnd_NestedTasks verifies that nested tasks correctly link parent IDs.
func TestEndToEnd_NestedTasks(t *testing.T) {
	tracker := newTestTracker(t, "nested.db")

	// Start parent.
	ctx, parent := tracker.StartTask(context.Background(), "workflow",
		core.WithCustomer("acme"),
	)

	// Start child (should auto-link parent).
	_, child := tracker.StartTask(ctx, "sub_step")

	if child.Task.ParentTaskID == nil {
		t.Fatal("expected parent_task_id on child")
	}
	if *child.Task.ParentTaskID != parent.Task.TaskID {
		t.Errorf("expected parent=%s, got %s", parent.Task.TaskID, *child.Task.ParentTaskID)
	}

	// End both.
	child.End(core.TaskStatusSuccess)
	parent.End(core.TaskStatusSuccess)

	// Verify in buffer.
	storedChild, _ := tracker.Buffer().GetTask(child.Task.TaskID.String())
	if storedChild.ParentTaskID == nil || *storedChild.ParentTaskID != parent.Task.TaskID {
		t.Error("stored child should have parent_task_id")
	}
}

// TestEndToEnd_FailedTask verifies that a failed task has failure_count=1.
func TestEndToEnd_FailedTask(t *testing.T) {
	tracker := newTestTracker(t, "failed.db")

	_, tt := tracker.StartTask(context.Background(), "failing_task")
	tt.RecordLLMCall("openai", "gpt-4o", 100, 50, core.WithCost(decimal.NewFromFloat(0.01)))
	tt.End(core.TaskStatusFailed)

	stored, _ := tracker.Buffer().GetTask(tt.Task.TaskID.String())
	if stored.FailureCount != 1 {
		t.Errorf("expected failure_count=1, got %d", stored.FailureCount)
	}
	if stored.Status != core.TaskStatusFailed {
		t.Errorf("expected failed, got %s", stored.Status)
	}
}

// TestEndToEnd_DecimalPrecision verifies that costs maintain full decimal precision
// through the entire pipeline.
func TestEndToEnd_DecimalPrecision(t *testing.T) {
	tracker := newTestTracker(t, "precision.db")

	_, tt := tracker.StartTask(context.Background(), "precision_test")
	preciseCost := decimal.RequireFromString("0.123456789012345678")
	tt.RecordCost("precise_svc", preciseCost)
	tt.End(core.TaskStatusSuccess)

	stored, _ := tracker.Buffer().GetTask(tt.Task.TaskID.String())
	if !stored.ExternalCostUSD.Equal(preciseCost) {
		t.Errorf("precision lost: expected %s, got %s", preciseCost, stored.ExternalCostUSD)
	}
}

// TestEndToEnd_SchemaVersion verifies schema_version is set on tasks and events.
func TestEndToEnd_SchemaVersion(t *testing.T) {
	tracker := newTestTracker(t, "schema.db")

	_, tt := tracker.StartTask(context.Background(), "schema_test")
	tt.RecordCost("svc", decimal.NewFromFloat(0.01))
	tt.End(core.TaskStatusSuccess)

	if tt.Task.SchemaVersion != "1" {
		t.Errorf("expected schema_version=1, got %s", tt.Task.SchemaVersion)
	}

	events, _ := tracker.Buffer().QueryEvents(tt.Task.TaskID.String())
	for _, e := range events {
		if e.SchemaVersion != "1" {
			t.Errorf("expected event schema_version=1, got %s", e.SchemaVersion)
		}
	}
}

File: rust/tests/integration_test.rs
use std::sync::Arc;

use dexcost::core::models::{Task, TaskStatus};
use dexcost::core::tracker::{TaskOptions, TrackedTask};
use dexcost::pricing::engine::PricingEngine;
use dexcost::transport::buffer::EventBuffer;
use rust_decimal::Decimal;
use tokio::sync::Mutex;

/// Full integration flow: create task -> record events -> end -> verify buffer.
#[tokio::test]
async fn test_full_flow() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));
    let pricing = Arc::new(Mutex::new(PricingEngine::new()));

    // Create a task with options
    let mut task = Task::new("resolve_ticket");
    task.customer_id = Some("acme-corp".to_string());
    task.project_id = Some("support".to_string());

    let mut tt = TrackedTask::new(task, buffer.clone(), Some(pricing));

    // Record an LLM call with explicit cost
    let llm_event = tt
        .record_llm_call(
            "openai",
            "gpt-4o",
            1000,
            500,
            Some(Decimal::new(5, 2)), // 0.05
            None,
            Some(250),
        )
        .await
        .unwrap();

    assert_eq!(llm_event.provider, Some("openai".to_string()));
    assert_eq!(llm_event.model, Some("gpt-4o".to_string()));

    // Record a non-LLM cost
    let cost_event = tt
        .record_cost("google_maps", Decimal::new(1, 2), None, None) // 0.01
        .await
        .unwrap();

    assert_eq!(cost_event.service_name, Some("google_maps".to_string()));

    // Mark a retry
    let retry_event = tt
        .mark_retry("rate_limit", Decimal::new(2, 2)) // 0.02
        .await
        .unwrap();

    assert!(retry_event.is_retry);

    // End the task
    tt.end(TaskStatus::Success).await.unwrap();

    // Verify task state
    let task = tt.task();
    assert_eq!(task.status, TaskStatus::Success);
    assert!(task.ended_at.is_some());
    assert_eq!(task.customer_id, Some("acme-corp".to_string()));
    assert_eq!(task.project_id, Some("support".to_string()));
    assert_eq!(task.llm_cost_usd, Decimal::new(5, 2));
    assert_eq!(task.external_cost_usd, Decimal::new(1, 2));
    assert_eq!(task.total_cost_usd, Decimal::new(6, 2)); // llm + external
    assert_eq!(task.retry_cost_usd, Decimal::new(2, 2));
    assert_eq!(task.retry_count, 1);
    assert_eq!(task.total_input_tokens, 1000);
    assert_eq!(task.total_output_tokens, 500);
    assert_eq!(task.schema_version, "1");

    // Verify buffer state
    let buf = buffer.lock().await;
    assert_eq!(buf.event_count(), 3); // llm + cost + retry
    assert_eq!(buf.task_count(), 1);
    assert!(buf.get_task(&task.task_id).is_some());
}

/// Test auto-pricing with the pricing engine.
#[tokio::test]
async fn test_auto_pricing_flow() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));
    let pricing = Arc::new(Mutex::new(PricingEngine::new()));

    let task = Task::new("auto_priced_task");
    let mut tt = TrackedTask::new(task, buffer.clone(), Some(pricing));

    // Record without explicit cost -- should auto-price
    let event = tt
        .record_llm_call("openai", "gpt-4o", 1000, 500, None, None, None)
        .await
        .unwrap();

    // If model is known, cost should be computed
    // If unknown, cost will be zero with unknown confidence
    // Either way, the event should be recorded
    assert!(tt.events().len() == 1);
    assert_eq!(event.event_type, dexcost::core::models::EventType::LlmCall);
}

/// Test that events end up in the buffer's pending list.
#[tokio::test]
async fn test_buffer_pending_after_records() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));
    let task = Task::new("buffer_test");
    let mut tt = TrackedTask::new(task, buffer.clone(), None);

    tt.record_llm_call(
        "openai",
        "gpt-4o",
        100,
        50,
        Some(Decimal::new(1, 2)),
        None,
        None,
    )
    .await
    .unwrap();
    tt.record_cost("svc", Decimal::new(2, 2), None, None)
        .await
        .unwrap();

    let buf = buffer.lock().await;
    let pending = buf.get_pending_events(100);
    assert_eq!(pending.len(), 2);
}

/// Test multiple tasks in the same buffer.
#[tokio::test]
async fn test_multiple_tasks_shared_buffer() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));

    let task1 = Task::new("task_type_a");
    let task1_id = task1.task_id.clone();
    let mut tt1 = TrackedTask::new(task1, buffer.clone(), None);

    let task2 = Task::new("task_type_b");
    let task2_id = task2.task_id.clone();
    let mut tt2 = TrackedTask::new(task2, buffer.clone(), None);

    tt1.record_llm_call(
        "openai",
        "gpt-4o",
        100,
        50,
        Some(Decimal::new(1, 2)),
        None,
        None,
    )
    .await
    .unwrap();
    tt2.record_cost("svc", Decimal::new(2, 2), None, None)
        .await
        .unwrap();

    tt1.end(TaskStatus::Success).await.unwrap();
    tt2.end(TaskStatus::Failed).await.unwrap();

    let buf = buffer.lock().await;
    assert_eq!(buf.event_count(), 2);
    assert_eq!(buf.task_count(), 2);
    assert!(buf.get_task(&task1_id).is_some());
    assert!(buf.get_task(&task2_id).is_some());
}

/// Test that to_dict produces valid JSON.
#[tokio::test]
async fn test_to_dict_json_output() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));
    let task = Task::new("json_test");
    let mut tt = TrackedTask::new(task, buffer.clone(), None);

    tt.record_llm_call(
        "openai",
        "gpt-4o",
        1000,
        500,
        Some(Decimal::new(5, 2)),
        None,
        None,
    )
    .await
    .unwrap();

    tt.end(TaskStatus::Success).await.unwrap();

    // Verify task to_dict produces valid JSON
    let task_dict = tt.task().to_dict();
    let json_str = serde_json::to_string(&task_dict).unwrap();
    assert!(!json_str.is_empty());

    // Verify event to_dict produces valid JSON
    for event in tt.events() {
        let event_dict = event.to_dict();
        let json_str = serde_json::to_string(&event_dict).unwrap();
        assert!(!json_str.is_empty());
    }
}

/// Test task options.
#[tokio::test]
async fn test_task_options() {
    let buffer = Arc::new(Mutex::new(EventBuffer::new().unwrap()));

    let mut task = Task::new("with_options");
    let opts = TaskOptions {
        customer_id: Some("customer-1".to_string()),
        project_id: Some("project-1".to_string()),
        experiment_id: Some("exp-1".to_string()),
        variant: Some("v2".to_string()),
        ..Default::default()
    };
    task.customer_id = opts.customer_id;
    task.project_id = opts.project_id;
    task.experiment_id = opts.experiment_id;
    task.variant = opts.variant;

    let mut tt = TrackedTask::new(task, buffer, None);
    tt.end(TaskStatus::Success).await.unwrap();

    assert_eq!(tt.task().customer_id, Some("customer-1".to_string()));
    assert_eq!(tt.task().project_id, Some("project-1".to_string()));
    assert_eq!(tt.task().experiment_id, Some("exp-1".to_string()));
    assert_eq!(tt.task().variant, Some("v2".to_string()));
}
